test/integration/diff_drive_system.cc

```diff
@@ -87,142 +75,175 @@ class Relay
-        msgs::Set(msg.mutable_linear(),
-                  math::Vector3d(std::min(nextLinVel, desiredLinVel), 0, 0));
-        msgs::Set(msg.mutable_angular(),
-                  math::Vector3d(0.0, 0, std::min(nextAngVel, desiredAngVel)));
-        pub.Publish(msg);
+        poses.push_back(poseComp->Data());
       });
+    server.AddSystem(testSystem.systemPtr);
 
-  server.AddSystem(velocityRamp.systemPtr);
-
-  server.Run(true, 3000, false);
+    // Run server and check that vehicle didn't move
+    server.Run(true, 1000, false);
 
-  // Poses for 4s
-  ASSERT_EQ(4000u, poses.size());
+    EXPECT_EQ(1000u, poses.size());
 
-  int sleep = 0;
-  int maxSleep = 30;
-  for (; odomPoses.size() < 150 && sleep < maxSleep; ++sleep)
-  {
-    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    for (const auto &pose : poses)
+    {
+      EXPECT_EQ(poses[0], pose);
+    }
+
+    // Get odometry messages
+    double period{1.0 / 50.0};
+    double lastMsgTime{1.0};
+    std::vector<math::Pose3d> odomPoses;
+    std::function<void(const msgs::Odometry &)> odomCb =
+      [&](const msgs::Odometry &_msg)
+      {
+        ASSERT_TRUE(_msg.has_header());
+        ASSERT_TRUE(_msg.header().has_stamp());
+
+        double msgTime =
+            static_cast<double>(_msg.header().stamp().sec()) +
+            static_cast<double>(_msg.header().stamp().nsec()) * 1e-9;
+
+        EXPECT_DOUBLE_EQ(msgTime, lastMsgTime + period);
+        lastMsgTime = msgTime;
+
+        odomPoses.push_back(msgs::Convert(_msg.pose()));
+      };
+
+    // Publish command and check that vehicle moved
+    transport::Node node;
+    auto pub = node.Advertise<msgs::Twist>(_cmdVelTopic);
+    node.Subscribe(_odomTopic, odomCb);
+
+    msgs::Twist msg;
+
+    // Avoid wheel slip by limiting acceleration
+    Relay velocityRamp;
+    const double desiredLinVel = 0.5;
+    const double desiredAngVel = 0.2;
+    const double linAccel = 1;
+    const double angAccel = 1;
+    velocityRamp.OnPreUpdate(
+        [&](const gazebo::UpdateInfo &_info,
+            const gazebo::EntityComponentManager &)
+        {
+          double dt = std::chrono::duration<double>(_info.dt).count();
+          double nextLinVel = msg.linear().x() + dt * linAccel;
+          double nextAngVel = msg.angular().z() + dt * angAccel;
+          msgs::Set(msg.mutable_linear(),
+              math::Vector3d(std::min(nextLinVel, desiredLinVel), 0, 0));
+          msgs::Set(msg.mutable_angular(),
+              math::Vector3d(0.0, 0, std::min(nextAngVel, desiredAngVel)));
+          pub.Publish(msg);
+        });
+
+    server.AddSystem(velocityRamp.systemPtr);
+
+    server.Run(true, 3000, false);
+
+    // Poses for 4s
+    ASSERT_EQ(4000u, poses.size());
+
+    int sleep = 0;
+    int maxSleep = 30;
+    for (; odomPoses.size() < 150 && sleep < maxSleep; ++sleep)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+    ASSERT_NE(maxSleep, sleep);
```

Comment:
The test refactoring looks good to me, I just think this could continue being an `EXPECT` because the test wouldn't crash if this fails.

Reply:
Yeah, I just didn't like the crash since it would block the other test in this file from running.

Reply:
Oh so if this line fails the test crashes?

Reply:
That's right.